MIDI-97: delete unnecessary tokens

WojciechMat opened the pull request:
I have deleted padding token from project logic. 
### bos and eos tokens
Removing \<s> and \</s> special tokens turned out to be more complicated than I anticipated. They play a valuable role in attention mechanism. The attention layer finds correlation between current and previous tokens, so existance of start token seem like necessity for now:

![image](https://github.com/Nospoko/midi-translation/assets/74838859/76135d26-9641-4b0e-a713-eb56ecec2c47)

When using greedy_decode we do need to pass starting token to the model at the beggining, so that attention mechanism has any sentence (composed of one token during first step) to work with. 

We could think of a way to overcome this and remove \<s> and \</s> tokens completly but I would have to think about it more as I do not see a viable solution right now. I will try to at least get rid of \</s> token now.

WojciechMat committed:
deleted padding token

roszcz commented:
I'm not convinced 😅 the chart you posted is from a language example, which does not apply in our case, because all of our "sentences" have exactly the same length (`sequence_len`), and we don't need additional tokens signal a change between sentences.

For the greedy decode, it should be possible to start with an empty sequence (attention_mask all negative?). If not, we just need to come up with some strategy for an initial velocity/dstart 🤔

WojciechMat committed:
fix batch mask for now

WojciechMat commented:
Starting with an empty sequence is not possibe, because we want to pass the starting sentence to https://github.com/Nospoko/midi-translation/blob/d80b24a43b6f76179d2cbe248aa7cabe34937e05/modules/decoder.py#L35
Which will return an empty tensor.
Unless of course I can think of a bypass.
I am not convinced as to starting velocity, because it will not depend on  the first src token at all - it does now. Unless we make it so :thinking:

WojciechMat committed:
remove '</s>' token, add torch.no_grad to dashboard, repair velocity dataset split

WojciechMat commented:
Deleted \</s> tokens. Thinking what to do with \<s>.

roszcz commented:
I see, let's rename it to "<CLS>" for now - this will be consistent with BERT approach. We can later try to use it to do composer classification or other downstream tasks

WojciechMat committed:
<s> -> ''

WojciechMat committed:
move src_mask generation to utils/greedy_decode and utils/decode_and_output

WojciechMat committed:
update generate_sequence

WojciechMat commented:
I moved src_mask generation in evaluations/dashboard to utils/greedy_decode and utils/decode_and_out, as they are always the same.
https://github.com/Nospoko/midi-translation/blob/338b05596740575d0c466ad2d2450af38749812c/utils.py#L114-L115

WojciechMat committed:
torch-like src_mask creation

WojciechMat committed:
src_mask to device

WojciechMat commented:
I moved src_mask creation to utils:
https://github.com/Nospoko/midi-translation/blob/bd9428c8579b704e99b5b94ca6b3c34ea1529896/utils.py#L140

and updated src_mask creation in Batch class, taking advantage of the fact we do not use any padding.

roszcz commented on data/tokenizer.py:
```diff
@@ -184,14 +174,10 @@ def quantize(self, start: list[float]) -> list[int]:
         # get tokens from quantized data
-        tokens = [str(dstart_bin) for dstart_bin in dstart_bins]
-        tokens = ["<s>"] + tokens + ["</s>"]
+        tokens = [""] + [str(dstart_bin) for dstart_bin in dstart_bins]
```
```suggestion
        tokens = [str(dstart_bin) for dstart_bin in dstart_bins]
```

Removing my comment about not adding tokens in this function does not address the issue 😅 

I think special tokens should be added in a step separated from tokenization of `dstart/velocity` bins, let me know if you see good reasons not to do this, because maybe I'm missing something.

Also, this first token shouldn't be an empty string, but rather a `"<CLS>" `token. This will make it consistent with other transformer-oriented research that we're doing.

WojciechMat replied:
Sorry, I deleted it when I was removing all the special tokens and thought it is resolved :sweat_smile: 

As for first token - I might have misunderstood your previeous comment - github hid the cls token :0.
> I see, let's rename it to "" for now - this will be consistent with BERT approach. We can later try to use it to do composer classification or other downstream tasks

WojciechMat committed:
use numel for counting items in tensor

WojciechMat commented:
How about I also delete doc.py and documentation folder as they were just a prototype for blog?

I changed the weird and obscure way of counting tokens for a simple tgt_y.numel().

WojciechMat committed:
"" -> "<CLS>"

WojciechMat committed:
add cls tokens in __getitem__ method of the dataset

WojciechMat commented:
cls tokens are now added inside \_\_getitem__ method of the dataset.

WojciechMat committed:
delete blog prototype

roszcz reviewed (approved):
🚀 

This is not related to changes in this PR, but I don't like the name `.untokenize` we use in encoders - let me know if you can come up with a better name for this 🤔

WojciechMat commented:
Perhabs we do not need this method at all - mayve we can replace it with MidiEncoder.decode and have it „untokenizing” directly from token ids

roszcz commented:
My priority here is to keep it as easy to read as possible - in terms of the data flow, I think we have a very clear transition `midi->pandas->quantized->tokens->token_ids` (and back). It would be great to always know which step and which direction is any procedure responsible for. 

With "untokenize" I can only guess it's the second half, going "back". "Decode" for me would be something like "tokens -> pandas". 🤔

WojciechMat commented:
What untokenize does is `tokens -> pandas` in QuantizedMidiEncoder or `tokens -> list[int]` in velocity and dstart encoders

We could rename it to decode or make decode take a list of `token_ids` and produce the same outcome.

There already is an implementation of decode in MidiEncoder, we just do not use it I think - it calls .untokenize after changing ids to tokens.

roszcz commented:
Let's merge it as is, recognizing and naming all the entities here is important, but it can wait 👍
